### PR TITLE
Initial README

### PR DESCRIPTION
**Pull request by claudiu-cc**
### Description

This PR adds a README file to the root of the repository and updates the godoc with some usage examples.

The structure of the README covers the following:
- Introduction: What is Encryptonize&reg; library and why would I use it?
- Features: What functionalities does the library offer and why would I use those functionalities?
- Installation: How do I install the library in my project?
- Usage: Examples of how to use the library functions.
- Security: Reference to the security architecture where I can see more details about the security parameters and guarantees offered by the library.

### Relevant Issues/PRs

Closes DEV-129
Closes DEV-130

### Type(s) of Change (Split the PR if you check many)

- [ ] Added user feature
- [ ] Added technical feature
- [ ] Added tests
- [ ] Refactor
- [ ] Bugfix
- [x] Updated documentation
- [ ] Kubernetes changes
- [ ] CI/CD workflow changes

### Testing Checklist

- [ ] I have added/updated unit tests for functional changes.
- [ ] I have added/updated integration tests for changes that affect dependent systems.
- [ ] I have added/updated end-to-end tests for feature changes.

### General Checklist

- [x] I have pulled in the latest changes from master.
- [ ] I have run `make lint`.
- [ ] I have successfully run `make tests`.
- [ ] I have updated relevant CI/CD workflows.
- [ ] I have updated relevant Kubernetes manifests/scripts.
- [ ] I have updated/added relevant (internal and external) documentation (readme, doc comments, etc).
- [ ] I have updated the dependency map to reflect my changes.
- [x] I have added the license to new source code files.
- [x] I have spent some time looking over the full diff before creating this PR.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+- **Discretionary Access Control**
+    - each encrypted object has individual access controls defined by the data owner
+    - data owners decide which users or user groups can access the decrypted objects
+- **User Authentication**
+    - authentication is done with custom credentials generated by the library functions
+- **Searcheable Symmetric Encryption (SSE)**
+    - the ability to efficiently search over encrypted objects without decrypting them can be added by creating an index of keywords and adding `(Keyword, ID)` pairs to it
+- **Encrypted Security Tokens**
+    - security tokens with built-in expiry times and arbitrary encrypted data can be generated and verified
+    - can be used for implementing token-based authentication/authorization schemes
```
Comment (philip-cc) [inline README.md:24]: Capitalisation and periods. 
```suggestion
- **Authenticated Encryption with Associated Data (AEAD)**
    - Binary data can be encrypted/decrypted using the standardized AES256-GCM algorithm.
    - The Ciphertext and Associated Data can be securely stored in untrusted locations as the recovery of the plaintext is computationally infeasible without the Encryption Key, and any changes to the Ciphertext or Associated Data will be detected during the decryption process.
- **Key Wrapping with Padding Authenticated Encryption (KWP-AE)**
    - Each object is encrypted with a random 256-bit key which is then wrapped with the configured Wrapping Key using the standardized KWP-AE algorithm.
    - Wrapped keys can be securely stored in untrusted locations (alongside the encrypted objects) as the recovery of the plaintext key is computationally infeasible without the Wrapping Key, and any changes to the Wrapped Key will be detected during the unwrapping process.
- **Discretionary Access Control**
    - Each encrypted object has individual access controls defined by the data owner.
    - Data owners decide which users or user groups can access the decrypted objects.
- **User Authentication**
    - Authentication is done with custom credentials generated by the library functions.
- **Searcheable Symmetric Encryption (SSE)**
    - The ability to efficiently search over encrypted objects without decrypting them can be added by creating an index of keywords and adding `(Keyword, ID)` pairs to it.
- **Encrypted Security Tokens**
    - Security tokens with built-in expiry times and arbitrary encrypted data can be generated and verified.
    - Can be used for implementing token-based authentication/authorization schemes.
```

File: README.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
+# Encryptonize&reg; library
+[![Library build](https://github.com/cyber-crypt-com/encryptonize-lib/actions/workflows/build.yaml/badge.svg)](https://github.com/cyber-crypt-com/encryptonize-lib/actions/workflows/build.yaml)
+[![Go Report Card](https://goreportcard.com/badge/github.com/cyber-crypt-com/encryptonize-lib)](https://goreportcard.com/report/github.com/cyber-crypt-com/encryptonize-lib)
+[![Go Reference](https://pkg.go.dev/badge/github.com/cyber-crypt-com/encryptonize-lib)](https://pkg.go.dev/github.com/cyber-crypt-com/encryptonize-lib)
+
+The Encryptonize&reg; library provides cryptographic functions for data encryption and decryption, as well as enforcing access control around the encrypted data. It allows developers to easily implement various security schemes and protocols for protecting data.
+
+## Features
+- **Authenticated Encryption with Associated Data (AEAD)**
+    - binary data can be encrypted/decrypted using the standardized AES256-GCM algorithm
```
Comment (philip-cc) [inline README.md:10]:
```suggestion
    - binary data can be encrypted/decrypted using the standardized AES256-GCM authenticated encryption algorithm
```

File: README.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    - the Ciphertext and Associated Data can be securely stored in untrusted locations as the recovery of the plaintext is computationally infeasible without the Encryption Key, and any changes to the Ciphertext or Associated Data will be detected during the decryption process
+- **Key Wrapping with Padding Authenticated Encryption (KWP-AE)**
+    - each object is encrypted with a random 256-bit key which is then wrapped with the configured Wrapping Key using the standardized KWP-AE algorithm
```
Comment (philip-cc) [inline README.md:13]: This refers to an "object", but we haven't defined what that is.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+[![Go Report Card](https://goreportcard.com/badge/github.com/cyber-crypt-com/encryptonize-lib)](https://goreportcard.com/report/github.com/cyber-crypt-com/encryptonize-lib)
+[![Go Reference](https://pkg.go.dev/badge/github.com/cyber-crypt-com/encryptonize-lib)](https://pkg.go.dev/github.com/cyber-crypt-com/encryptonize-lib)
+
+The Encryptonize&reg; library provides cryptographic functions for data encryption and decryption, as well as enforcing access control around the encrypted data. It allows developers to easily implement various security schemes and protocols for protecting data.
+
+## Features
+- **Authenticated Encryption with Associated Data (AEAD)**
+    - binary data can be encrypted/decrypted using the standardized AES256-GCM algorithm
+    - the Ciphertext and Associated Data can be securely stored in untrusted locations as the recovery of the plaintext is computationally infeasible without the Encryption Key, and any changes to the Ciphertext or Associated Data will be detected during the decryption process
+- **Key Wrapping with Padding Authenticated Encryption (KWP-AE)**
```
Comment (philip-cc) [inline README.md:12]:
```suggestion
- **Key Wrapping with Authenticated Encryption (KWP-AE)**
```

Comment (philip-cc) [inline README.md:12]: KWP-AE is a algorithm, whereas AEAD in the headline above is a concept.

File: README.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    - each object is encrypted with a random 256-bit key which is then wrapped with the configured Wrapping Key using the standardized KWP-AE algorithm
+    - wrapped keys can be securely stored in untrusted locations (alongside the encrypted objects) as the recovery of the plaintext key is computationally infeasible without the Wrapping Key, and any changes to the Wrapped Key will be detected during the unwrapping process
+- **Discretionary Access Control**
+    - each encrypted object has individual access controls defined by the data owner
+    - data owners decide which users or user groups can access the decrypted objects
+- **User Authentication**
+    - authentication is done with custom credentials generated by the library functions
+- **Searcheable Symmetric Encryption (SSE)**
```
Comment (philip-cc) [inline README.md:20]: We should probably mention what the security gains are for this feature.

File: README.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    - the ability to efficiently search over encrypted objects without decrypting them can be added by creating an index of keywords and adding `(Keyword, ID)` pairs to it
+- **Encrypted Security Tokens**
+    - security tokens with built-in expiry times and arbitrary encrypted data can be generated and verified
+    - can be used for implementing token-based authentication/authorization schemes
+
+## Installation
+To use the library in your Go project, you first need to install it by running the following command in the root folder of the project:
```
Comment (philip-cc) [inline README.md:27]:
```suggestion
To use the library in your Go project, you first need to add it to your `go.mod` file by running the following command in the root folder of the project:
```

File: README.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+go get -u github.com/cyber-crypt-com/encryptonize-lib@latest
+```
+
+Then, you can access the Encryptonize&reg; library functions by importing the library in your application code:
+```
+import github.com/cyber-crypt-com/encryptonize-lib
+```
+
+## Usage
+For examples on how to use the Encryptonize&reg; library [see our Examples section in the documentation](https://pkg.go.dev/github.com/cyber-crypt-com/encryptonize-lib#example-package-BasicEncryptDecrypt).
```
Comment (philip-cc) [inline README.md:38]: We should wait with this link until the license issue is fixed. A TODO is fine.

File: README.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
+# Encryptonize&reg; library
+[![Library build](https://github.com/cyber-crypt-com/encryptonize-lib/actions/workflows/build.yaml/badge.svg)](https://github.com/cyber-crypt-com/encryptonize-lib/actions/workflows/build.yaml)
+[![Go Report Card](https://goreportcard.com/badge/github.com/cyber-crypt-com/encryptonize-lib)](https://goreportcard.com/report/github.com/cyber-crypt-com/encryptonize-lib)
+[![Go Reference](https://pkg.go.dev/badge/github.com/cyber-crypt-com/encryptonize-lib)](https://pkg.go.dev/github.com/cyber-crypt-com/encryptonize-lib)
+
+The Encryptonize&reg; library provides cryptographic functions for data encryption and decryption, as well as enforcing access control around the encrypted data. It allows developers to easily implement various security schemes and protocols for protecting data.
+
+## Features
```
Comment (philip-cc) [inline README.md:8]: I think we should discuss the aim of this section.

File: example_basic_test.go [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+// This is a basic example demonstrating how to use the Encryptonize® library to encrypt and decrypt binary data.
+func Example_basicEncryptDecrypt() {
+	// Instantiate the Encryptonize® library with the given keys.
+	ectnz, err := encryptonize.New(keys)
+	if err != nil {
+		log.Fatalf("Error instantiating Encryptonize: %v", err)
+	}
+
+	// Create a basic user.
+	ectnzUser, _, _, err := ectnz.NewUser(nil)
```
Comment (philip-cc) [inline example_basic_test.go:43]: Let's keep it simple. 
```suggestion
	user, _, _, err := ectnz.NewUser(nil)
```

File: example_token_test.go [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+package encryptonize_test
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/cyber-crypt-com/encryptonize-lib"
+)
+
+func ExampleEncryptonize_CreateToken() {
```
Comment (philip-cc) [inline example_token_test.go:24]: There should probably be some more explanatory comments here to match the other examples.

File: README.md [hunk 11]
```diff
@@ -0,0 +1,41 @@
+# Encryptonize&reg; library
+[![Library build](https://github.com/cyber-crypt-com/encryptonize-lib/actions/workflows/build.yaml/badge.svg)](https://github.com/cyber-crypt-com/encryptonize-lib/actions/workflows/build.yaml)
+[![Go Report Card](https://goreportcard.com/badge/github.com/cyber-crypt-com/encryptonize-lib)](https://goreportcard.com/report/github.com/cyber-crypt-com/encryptonize-lib)
+[![Go Reference](https://pkg.go.dev/badge/github.com/cyber-crypt-com/encryptonize-lib)](https://pkg.go.dev/github.com/cyber-crypt-com/encryptonize-lib)
```
[the remark this answers is not in the document]
Comment (claudiu-cc) [reply README.md]: Fixed

File: README.md (hunk 2 above)
Comment (claudiu-cc) [reply README.md]: Fixed

File: README.md (hunk 3 above)
Comment (claudiu-cc) [reply README.md]: Fixed

File: README.md (hunk 4 above)
Comment (claudiu-cc) [reply README.md]: Fixed

Comment (claudiu-cc) [reply README.md]: Fixed. I left only the concepts in the subtitles.

File: README.md (hunk 6 above)
Comment (claudiu-cc) [reply README.md]: Fixed

File: README.md (hunk 7 above)
Comment (claudiu-cc) [reply README.md]: Fixed

File: example_basic_test.go (hunk 9 above)
Comment (claudiu-cc) [reply example_basic_test.go]: Fixed

File: example_token_test.go (hunk 10 above)
Comment (claudiu-cc) [reply example_token_test.go:24]: Fixed

File: example_access_control_test.go [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+type PublicUserData struct {
+	group               encryptonize.SealedGroup
+	encryptedData       encryptonize.SealedObject
+	encryptedDataAccess encryptonize.SealedAccess
+}
+
+// createUser instantiates a user with its private and public data.
+func createUser(ectnz encryptonize.Encryptonize) UserData {
```
Comment (philip-cc) [inline example_access_control_test.go:44]: Wouldn't `createUserData` make more sense?

File: example_access_control_test.go [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+	// Instantiate the Encryptonize® library with the given keys.
+	ectnz, err := encryptonize.New(keys)
+	if err != nil {
+		log.Fatalf("Error instantiating Encryptonize: %v", err)
+	}
+
+	// Create three users with their data.
+	user1 := createUser(ectnz)
+	user2 := createUser(ectnz)
+	user3 := createUser(ectnz)
```
Comment (philip-cc) [inline example_access_control_test.go:77]: Could we maybe call them something more distinct, like "alice", "bob", and "charlie"?

File: example_access_control_test.go (hunk 12 above)
Comment (claudiu-cc) [reply example_access_control_test.go]: Fixed

File: example_access_control_test.go (hunk 13 above)
Comment (claudiu-cc) [reply example_access_control_test.go]: Fixed

File: README.md (hunk 5 above)
Comment (claudiu-cc) [reply README.md:17]: Fixed

**Review by philip-cc — APPROVED** [review]

**Pull request merged**